### PR TITLE
Add docker support 

### PR DESCRIPTION
**Pull request by wyegelwel**
This code was derived from sdegler's pull request #91.

### CODE REVIEW

File: scheduler/src/cook/mesos/api.clj [hunk 1]
```diff
@@ -58,15 +86,76 @@
...
+                                {:docker.param/key key
+                                 :docker.param/value value})
+                              params)}))
+
+(defn- mkvolumes
+  "Helper for build-container.  Transforms volumes into the datomic schema."
+  [cid vols]
+  (when-not (empty? vols)
+    (let [vol-maps (mapv (fn [{:keys [container-path host-path mode]}]
+                           (merge (when container-path
```
Comment (leifwalsh) [inline scheduler/src/cook/mesos/api.clj]: (into {} (filter second {:container.volume/container-path container-path
                             :container.volume/host-path host-path
                             :container.volume/mode (clojure.string/upper-case mode)}))

Comment (leifwalsh) [inline scheduler/src/cook/mesos/api.clj]: (only if `(nil? (clojure.string/upper-case nil))`)

File: scheduler/src/cook/mesos/api.clj [hunk 2]
```diff
@@ -58,15 +86,76 @@
...
+                                     {:docker.portmap/host-port host-port})
+                                   (when protocol
+                                     {:docker.portmap/protocol (clojure.string/upper-case protocol)})))
+                          ports)]
+      {:docker/port-mapping port-maps})))
+
+(defn- build-container
+  "Helper for submit-jobs, deal with container structure."
+  [id container]
+  (let [container-id (d/tempid :db.part/user)
```
Comment (leifwalsh) [inline scheduler/src/cook/mesos/api.clj]: (when (and container (= (clojure.string/lower-case (:type container)) "docker))
      (let ...))

File: scheduler/src/cook/mesos/api.clj [hunk 3]
```diff
@@ -103,6 +192,7 @@
                                   :label/key k
                                   :label/value v}]))
                             labels)
+                container (if (nil? container) [] (build-container id container))
```
Comment (leifwalsh) [inline scheduler/src/cook/mesos/api.clj]: (or (build-container id container) [])

File: scheduler/src/cook/mesos/api.clj (hunk 1 above)
Comment (leifwalsh) [inline scheduler/src/cook/mesos/api.clj]: (also, optional, I'm not sure which is better)

**Comment by leifwalsh** [comment]
LGTM

File: scheduler/src/cook/mesos/api.clj (hunk 1 above)
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]:
```clojure
user=> (clojure.string/upper-case nil)

NullPointerException   clojure.string/upper-case (string.clj:209)
````

File: scheduler/src/cook/mesos/api.clj [hunk 4]
```diff
@@ -58,15 +86,76 @@
...
+(defn- mk-container-params
+  "Helper for build-container.  Transforms parameters into the datomic schema."
+  [cid params]
+  (when-not (empty? params)
+    {:docker/parameters (mapv (fn [{:keys [key value]}]
+                                {:docker.param/key key
+                                 :docker.param/value value})
+                              params)}))
+
+(defn- mkvolumes
```
Comment (icexelloss) [inline scheduler/src/cook/mesos/api.clj]: mk-volumns? (Not important)

**Comment by wyegelwel** [comment]
Will make another PR to address cosmetic concerns

**Pull request merged**